### PR TITLE
[Feat] : Add AdaGrad, RMSProp, and Adam Optimizers

### PR DESCRIPTION
**Pull request by Advaitgaur004**
Just added a few more optimisers to have some more powerful options for training! This PR brings in AdaGrad, RMSProp, and the classic Adam.

- Added the new optimizers (AdaGrad, RMSProp, Adam) to the main optimizer file. I put them all in src/optimizer/sgd.c for now, but we can totally break them into separate files later if we want.
- Updated cten.h with the new function declarations so they can be used across the project.
- Kept SGD as is, so nothing breaks there.

### CODE REVIEW

**Comment by Advaitgaur004** [comment]
to test the implemenation use this ==> https://gist.github.com/Advaitgaur004/e0ad0e99f1e4514bd6091f3cce10b23a

**Comment by Advaitgaur004** [comment]
image
 
As expected, AdaGrad diminishes the learning rate, and after 300 iterations, it is about to reach the destination, but not entirely.

**Comment by Advaitgaur004** [comment]
Choosing an optimizer can be tricky....many people default to Adam, but sometimes SGDM performs better.
Refer to this --> [Here](https://arxiv.org/abs/2010.05627)
image

File: src/optimizer/sgd.c
Comment (Advaitgaur004) [inline src/optimizer/sgd.c:1]: Extracted symbols from: https://gist.github.com/mrquincle/1102e58562411b7f633c08d5dc9a2e1f
